core/src/bitcoin_syncer.rs

```diff
@@ -0,0 +1,205 @@
+    Ok(BlockInfo {
+        block_hash,
+        block_header,
+        block_height: height,
+    })
+}
+
+pub async fn process_block(
+    db: &Database,
+    block: &bitcoin::Block,
+    block_height: i64,
+) -> Result<(), BridgeError> {
+    let mut dbtx = db.begin_transaction().await?;
+    let block_hash = block.header.block_hash();
+
+    db.set_chain_head(
+        Some(&mut dbtx),
+        &block_hash,
+        &block.header.prev_blockhash,
+        block_height,
+    )
+    .await?;
+
+    for tx in &block.txdata {
+        process_tx(db, &mut dbtx, tx, &block_hash).await?;
+    }
+
+    dbtx.commit().await?;
+    Ok::<(), BridgeError>(())
+}
+
+pub async fn process_tx(
+    db: &Database,
+    dbtx: DatabaseTransaction<'_, '_>,
+    tx: &bitcoin::Transaction,
+    block_hash: &bitcoin::BlockHash,
+) -> Result<(), BridgeError> {
+    let txid = tx.compute_txid();
+    db.insert_tx(dbtx, block_hash, &txid).await?;
+
+    for input in &tx.input {
+        db.insert_utxo(
+            dbtx,
+            &txid,
+            &input.previous_output.txid,
+            input.previous_output.vout as i64,
+        )
+        .await?;
+    }
+
+    Ok::<(), BridgeError>(())
+}
+
+pub async fn start_bitcoin_syncer(
+    db: Database,
+    rpc: ExtendedRpc,
+    poll_delay: Duration,
+    mode: BitcoinSyncerPollingMode,
+) -> Result<ChainHeadPollingResult, BridgeError> {
+    let (tx, _) = broadcast::channel(100);
+
+    let returned_tx = tx.clone();
+
+    let mut block_height = db
+        .get_max_height(None)
+        .await?
+        .ok_or(BridgeError::BlockNotFound)?;
+
+    let handle = tokio::spawn(async move {
+        loop {
+            let mut block_hash = rpc.client.get_block_hash(block_height + 1).await?;
+            let mut block_header = rpc.client.get_block_header(&block_hash).await?;
+
+            let mut new_blocks = vec![BlockInfo {
+                block_hash,
+                block_header,
+                block_height: block_height + 1,
+            }];
+
+            for _ in 0..100 {
+                // if the previous block is in the db, do nothing
+                let height = db
+                    .get_height_from_block_hash(None, block_header.prev_blockhash)
+                    .await?;
+                if height.is_some() {
+                    break;
+                }
+
+                // if the previous block is not in the db, we need to get the previous block
+
+                block_hash = block_header.prev_blockhash;
+                block_header = rpc.client.get_block_header(&block_hash).await?;
+
+                let block_info = BlockInfo {
+                    block_hash,
+                    block_header,
+                    block_height,
+                };
+                new_blocks.push(block_info);
+
+                block_height -= 1;
+            }
+
+            // If we haven't found a match after 100 blocks, the database is too far out of sync
+            if new_blocks.len() == 100 {
```

Comment:
duplicated check, let's just make the above `for` a `loop` then break using this `if`